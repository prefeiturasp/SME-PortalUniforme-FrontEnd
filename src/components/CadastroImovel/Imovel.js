import React, { Component } from "react";

import { Field } from "redux-form";
import { InputText } from "components/Input/InputText";
import { required } from "helpers/fieldValidators";

import { FileUpload } from "components/Input/FileUpload";
import { AutoComplete } from "components/Input/AutoComplete";
import { Card } from "primereact/card";
import { API_IMOVEIS_DEMANDA } from "constants/apiFilaCreche.constants";

export class Imovel extends Component {
  constructor(props) {
    super(props);
    this.state = {
      endereco: "",
      bairro: "",
      selectCEP: [],
      cep: "",
      helpText: ""
    };
    this.grupo = [
      [1, "Bercario I"],
      [4, "Bercario II"],
      [27, "Mini Grupo I"],
      [28, "Mini grupo II"]
    ];

    this.handleAddressChange = this.handleAddressChange.bind(this);
    this.onFetchCEP = this.onFetchCEP.bind(this);
    this.onFetchGrupoCreche = this.onFetchGrupoCreche.bind(this);
    this.onAddressBlur = this.onAddressBlur.bind(this);
  }

  onFetchCEP = dataAddress => {
    this.onFetchGrupoCreche(dataAddress);
  };

  onFetchGrupoCreche(endereco) {
    const authToken = {
      Authorization: `Token 4fcce3bb12805893514061555474d867adc85329`,
      "Content-Type": "application/json"
    };
    const { longitude, latitude } = endereco;
    fetch(`${API_IMOVEIS_DEMANDA}/${latitude}/${longitude}`, {
      headers: authToken
    })
      .then(response => response.json())
      .then(json => {
        const grupo = this.grupo;
        let data = {};
        grupo.forEach((faixa, key) => {
          let faixaCorrespondente = json.results.find(
            result => result["cd_serie_ensino"] === faixa[0]
          );
          data[faixa[0]] = faixaCorrespondente ? faixaCorrespondente.total : 0;
        });
        this.setState(data);
      });
  }

  handleAddressChange(dataAddress) {
    if (dataAddress.cep) {
      this.setState({
        helpText: ""
      });
    }
    this.onFetchCEP(dataAddress);
    this.setState({
      ...dataAddress
    });
    this.props.setAddressSelected(true, dataAddress);
  }

  onAddressBlur(event) {
    let helpText = "";
    if (this.state.cep === "") {
      helpText = "Endereço inválido. Selecione um resultado da lista.";
    } 
    if (this.state.numero === undefined) {
      helpText = "É necessário informar o número do imóvel no endereço.";
    }
    this.setState({ helpText });
  }

  render() {
    const { AddressSelected } = this.props;
    const { bairro, endereco, helpText } = this.state;
    return (
      <div className="card">
        <div className="card-body">
          <h4 className="card-title">Dados do Imóvel</h4>

          <div className="row">
            <div className="p-col-12 p-md-6">
              <Field
                component={AutoComplete}
                label="Endereço"
                name="endereco.endereco"
                required
                validate={required}
                handleChange={this.handleAddressChange}
                onAddressBlur={this.onAddressBlur}
                helpText={helpText}
                {...this.props}
              />
              <div className="row">
                <div className="col-12">
                  <Field
                    component={InputText}
                    label="Complemento"
                    name="endereco.complemento"
                  />
                </div>
              </div>

              <Field
                component={InputText}
                label="Número do Contribuinte (IPTU)"
                name="endereco.numero_iptu"
              />

              <Field
                component={FileUpload}
                name="planta_fotos"
                id="planta_fotos"
                accept="file/pdf"
                className="form-control-file"
                label="Fotos / Planta Baixa"
                required
                validate={required}
                {...this.props}
              />
            </div>
            <div className="p-col-12 p-md-6">
              {AddressSelected && (
                <div>
                  <Card title="Endereço Selecionado">
                    <b>Endereço:</b> <span>{endereco.replace(', São Paulo, Brasil', '')}</span>
                    <br />
                    <b>Bairro:</b> <span>{bairro}</span>
                    <br />
                    <b>Cidade:</b> <span>São Paulo </span>
                    <b>Estado:</b> <span>SP</span>
                  </Card>

                  <Card title="Demanda da Região">
                    <table className="table table-striped">
                      <thead>
                        <tr>
                          <th scope="col">Grupo</th>
                          <th scope="col">Fila</th>
                        </tr>
                      </thead>
                      <tbody>
                        {this.grupo.map((item, i) => {
                          return (
                            <tr key={i}>
                              <td>{item[1]}</td>
                              <td>{this.state[item[0]]}</td>
                            </tr>
                          );
                        })}
                      </tbody>
                    </table>
                  </Card>
                </div>
              )}
            </div>
          </div>
        </div>
      </div>
    );
  }
}
