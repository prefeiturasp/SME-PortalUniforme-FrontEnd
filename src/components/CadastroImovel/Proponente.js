import React, { Component } from "react";

import { Field } from "redux-form";
import { InputText } from "components/Input/InputText";
import { SelectText } from "components/Input/SelectText";
import { fieldCPF_CNPJ, fieldTel } from "helpers/textMask";
import { email } from "helpers/fieldValidators";
import { TIPO_PROPONENTE } from "constants/choices.constants";


export class Proponente extends Component {
  render() {
    return (
      <div className="card">
        <div className="card-body">
          <h4 className="card-title">Proponente</h4>
          <Field
            component={SelectText}
            options={TIPO_PROPONENTE}
            placeholder="Selecione um Tipo"
            label="Tipo"
            name="proponente.tipo"
          />
          <Field component={InputText} label="Nome" name="proponente.nome" />
          <Field
            customChange={fieldCPF_CNPJ}
            component={InputText}
            label="CPF/CNPJ"
            name="proponente.cpf_cnpj"
            maxLength="18"
          />
          <Field
            component={InputText}
            label="E-mail"
            name="proponente.email"
            validate={email}
          />
          <Field
            {...fieldTel}
            component={InputText}
            label="Telefone"
            name="proponente.telefone"
          />
        </div>
      </div>
    );
  }
}
