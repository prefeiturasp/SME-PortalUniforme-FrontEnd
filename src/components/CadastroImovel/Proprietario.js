import React, { Component } from "react";
import { Field } from "redux-form";
import { InputText } from "components/Input/InputText";
import { fieldTel, fieldCPF_CNPJ } from "helpers/textMask";
import { required, email } from "helpers/fieldValidators";

export class Proprietario extends Component {
  render() {
    return (
      <div className="card">
        <div className="card-body">
          <h4 className="card-title">Proprietário do Imóvel</h4>
          <Field
            component={InputText}
            label="Nome"
            name="contato.nome"
            required
            validate={required}
            autoFocus
          />
          <Field
            customChange={fieldCPF_CNPJ}
            component={InputText}
            label="CPF/CNPJ"
            name="contato.cpf_cnpj"
            required
            validate={required}
            maxLength="18"
          />
          <Field
            component={InputText}
            label="E-mail"
            name="contato.email"
            required
            validate={required && email}
          />
          <Field
            {...fieldTel}
            component={InputText}
            label="Telefone"
            name="contato.telefone"
            required
            validate={required}
          />
        </div>
      </div>
    );
  }
}
