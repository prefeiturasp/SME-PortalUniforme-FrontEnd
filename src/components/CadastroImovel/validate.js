import { validarCNPJ, validarCPF, hasNumber } from "helpers/utils";

export const validarForm = values => {
  let erro = null;
  if (
    !validarCNPJ(values.contato.cpf_cnpj) &&
    !validarCPF(values.contato.cpf_cnpj)
  ) {
    erro = "CPF ou CNPJ do Proprietário inválido";
  } else if (values.proponente && !hasNumber(values.proponente.telefone)) {
    delete values.proponente.telefone;
  }
  return erro;
};
