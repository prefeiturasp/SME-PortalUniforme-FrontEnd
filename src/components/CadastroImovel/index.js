import React, { Component } from "react";
import { Form, reduxForm } from "redux-form";
import { Messages } from "primereact/messages";

import BasePage from "components/Base/BasePage";
import { Imovel as ImovelService } from "services/Imovel.service";

import { Proponente } from "./Proponente";
import { Proprietario } from "./Proprietario";
import { Imovel } from "./Imovel";

// Style PrimeReact
import "primereact/resources/themes/nova-light/theme.css";
import "primereact/resources/primereact.min.css";
import "primeicons/primeicons.css";
import "primeflex/primeflex.css";
import "../../styles/styles.scss";
import "bootstrap/dist/css/bootstrap.min.css";
import { ModalConfirmacaoCadastro } from "./components/ModalConfirmacaoCadastro";
import { toastError } from "components/Toast/dialogs";
import { getError } from "helpers/utils";
import { validarForm } from "./validate";

const ENTER = 13;

export class CadastroImovel extends Component {
  constructor() {
    super();
    this.state = {
      imageFile: [],
      AddressSelected: false,
      endereco: "",
      bairro: "",
      latitude: "",
      longitude: "",
      selectCEP: [],
      cep: "",
      numero: "",
      resetarFile: false,
      protocolo: null,
      showModal: false,
      labelBotao: "Enviar"
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onKeyPress = this.onKeyPress.bind(this);
    this.setAddressSelected = this.setAddressSelected.bind(this);
    this.closeModal = this.closeModal.bind(this);
    this.setResetarFileFalse = this.setResetarFileFalse.bind(this);
  }

  onKeyPress(event) {
    if (event.which === ENTER) {
      event.preventDefault();
    }
  }

  closeModal() {
    this.setState({ showModal: false });
  }

  onSubmit(values) {
    const erro = validarForm(values);
    if (erro) {
      toastError(erro);
    } else {
      values.endereco.endereco = this.state.endereco;
      values.endereco.cep = this.state.cep;
      values.endereco.bairro = this.state.bairro;
      values.endereco.numero = this.state.numero;
      values.endereco.longitude = this.state.longitude;
      values.endereco.latitude = this.state.latitude;
      values.contato.telefone = values.contato.telefone.replace("_", "");
      if (values.proponente && values.proponente.telefone) {
        values.proponente.telefone = values.proponente.telefone.replace(
          "_",
          ""
        );
      }
      values["endereco"] = { ...values.endereco };
      this.setState({ labelBotao: "Aguarde..." });
      ImovelService.create(values)
        .then(resp => {
          if (resp.status === 201) {
            this.resetForm();
            this.setState({
              showModal: true,
              protocolo: resp.data.protocolo,
              labelBotao: "Enviar"
            });
          } else if (resp.status === 400) {
            toastError(getError(resp.data));
            this.setState({ labelBotao: "Enviar" });
          } else if (resp.status === 413) {
            toastError("O tamanho total máximo dos arquivos é 10MB");
            this.setState({ labelBotao: "Enviar" });
          } else {
            toastError("Erro ao efetuar cadastro de Imóvel");
            this.setState({ labelBotao: "Enviar" });
          }
        })
        .catch(error => {
          toastError(getError(error));
          this.setState({ labelBotao: "Enviar" });
        });
    }
  }

  resetForm = () => {
    this.setState({
      AddressSelected: false,
      resetarFile: true
    });
    this.props.reset();
  };

  setResetarFileFalse() {
    this.setState({ resetarFile: false });
  }

  setAddressSelected(value, address) {
    this.setState({
      AddressSelected: value,
      cep: address.cep,
      bairro: address.bairro,
      numero: address.numero,
      endereco: address.endereco,
      latitude: address.latitude,
      longitude: address.longitude
    });
  }

  render() {
    const { labelBotao, protocolo, showModal } = this.state;
    const { handleSubmit, pristine, submitting } = this.props;
    return (
      <BasePage>
        <ModalConfirmacaoCadastro
          protocolo={protocolo}
          showModal={showModal}
          closeModal={this.closeModal}
        />
        <div id="conteudo" className="w-100 desenvolvimento-escolar">
          <div className="container pt-5 pb-5">
            <div className="row">
              <div className="col-12 mb-4">
                <h1>Cadastro de Oferta de Imóvel</h1>
              </div>
              <div className="col-12">
                <Messages ref={el => (this.messages = el)}></Messages>
              </div>
              <div className="col-12">
                <Form
                  onSubmit={handleSubmit(this.onSubmit)}
                  onKeyPress={this.onKeyPress}
                  className="row"
                >
                  {/* Proponente */}
                  <div className="p-col-12 p-md-6">
                    <Proponente />
                  </div>

                  {/* Proprietário */}
                  <div className="p-col-12 p-md-6">
                    <Proprietario />
                  </div>

                  {/* Imovel */}
                  <div className="p-col-12">
                    <Imovel
                      {...this.state}
                      setResetarFileFalse={this.setResetarFileFalse}
                      setAddressSelected={this.setAddressSelected}
                    />
                  </div>

                  {/* Botao */}
                  <div className="botoes">
                    <button
                      type="button"
                      className="btn btn-outline-danger"
                      disabled={pristine || submitting}
                      onClick={this.resetForm}
                    >
                      Limpar
                    </button>
                    <button
                      type="submit"
                      className="btn btn-primary"
                      disabled={
                        pristine || submitting || labelBotao === "Aguarde..."
                      }
                    >
                      {labelBotao}
                    </button>
                  </div>
                </Form>
              </div>
            </div>
          </div>
        </div>
      </BasePage>
    );
  }
}

CadastroImovel = reduxForm({
  // a unique name for the form
  form: "CadastroImovelForm"
})(CadastroImovel);

export default CadastroImovel;
