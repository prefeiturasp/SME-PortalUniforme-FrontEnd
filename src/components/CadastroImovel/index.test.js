import { shallow } from "enzyme";
import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

import CadastroImovel from ".";

describe("test <CadastroImovel>", () => {
  const mockStore = configureStore();
  const initialState = {};

  let wrapper;
  beforeAll(() => {
    let store = mockStore(initialState);

    wrapper = shallow(
      <Provider store={store}>
        <CadastroImovel />
      </Provider>
    );
  });

  it("renders component", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
