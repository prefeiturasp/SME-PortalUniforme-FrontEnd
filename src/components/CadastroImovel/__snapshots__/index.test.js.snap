// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <CadastroImovel> renders component 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={false}
    forceUnregisterOnUnmount={false}
    form="CadastroImovelForm"
    getFormState={[Function]}
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
  />
</ContextProvider>
`;
